File: jupyter2/problems/recursive_type/binheap/go/binheap.go
package main

import (
	"fmt"
	"slices"
)

/** begin my answer */
/** end my answer */

// add all values in list xs to tree t in order
func addSeq(xs []int, t *BinHeap) *BinHeap {
	for _, x := range xs {
		t = binHeapAdd(x, t)
	}
	return t
}

// remove all values from t and push them in front of xs
func binHeapToSeq(t *BinHeap) []int {
	xs := append([]int{})
	for t != nil {
		x, t_ := binHeapRemoveMin(t)
		xs = append(xs, x)
		t = t_
	}
	return xs
}

// 0 ... n - 1
func mkRange(n int) []int {
	s := make([]int, n)
	for i := 0; i < n; i++ {
		s[i] = i
	}
	return s
}

// random sequence
func randomSeq(n int, x int) []int {
	seq := make([]int, n)
	for i := 0; i < n; i++ {
		x = (0x5DEECE66D * x + 0xB) & 0xFFFFFFFFFFFF
		seq[i] = x >> 17
	}
	return seq
}

// randomly shuffle xs
type Pair struct {
	k int
	x int
}

func randomShuffle(seed int, xs []int) []int {
	ys := randomSeq(len(xs), seed)
	ps := make([]Pair, len(xs))
	for i, x := range xs {
		ps[i].k = ys[i]
		ps[i].x = x
	}
	slices.SortFunc(ps, func(a, b Pair) int {
		return a.k - b.k
	})
	for i, z := range ps {
		ys[i] = z.x
	}
	return ys
}

// add xs to empty tree; remove all elements from the tree;
// check if they are sorted

func printSlice(header string, xs []int, n int) {
	fmt.Print(header, "[")
	for i, x := range xs {
		if i >= n {
			fmt.Print(" ...")
			break
		}
		if i > 0 {
			fmt.Print(", ")
		}
		fmt.Print(x)
	}
	fmt.Println("]")	
}	

func checkSeq(xs []int) bool {
	printSlice("input:  ", xs, 7)
	t := addSeq(xs, nil)
	ys := binHeapToSeq(t)
	printSlice("output: ", ys, 7)
	zs := slices.Sorted(slices.Values(xs))
	return slices.Equal(ys, zs)
}

// randomly shuffle 0 .. n-1;
//   add them to empty tree;
//   remove all elements from the tree;
//   check they are sorted

func checkRandom(seed int, n int) bool {
	rng := mkRange(n)
	xs := randomShuffle(seed, rng)
	return checkSeq(xs)
}

func main() {
	seed := 12345
	if (!checkRandom(seed, 2))      { panic("wrong") }
	if (!checkRandom(seed, 3))      { panic("wrong") }
	if (!checkRandom(seed, 4))      { panic("wrong") }
	if (!checkRandom(seed, 5))      { panic("wrong") }
	if (!checkRandom(seed, 10))     { panic("wrong") }
	if (!checkRandom(seed, 100))    { panic("wrong") }
	if (!checkRandom(seed, 1000))   { panic("wrong") }
	if (!checkRandom(seed, 10000))  { panic("wrong") }
	if (!checkRandom(seed, 100000)) { panic("wrong") }
	fmt.Println("OK")
}

File: jupyter2/problems/recursive_type/binheap/ml/binheap.ml
(** begin my answer *)

(** end my answer *)

(* add all values in list xs to tree t in order *)
let rec add_seq xs t =
  match xs with
    [] -> t
  | x :: rs -> add_seq rs (binheap_add x t)
;;

(* remove all values from t and push them into xs *)
let rec binheap_to_seq_ t xs =
  match t with
    Empty -> xs
  | _ ->
     let x, t' = binheap_remove_min t in
     binheap_to_seq_ t' (x :: xs)
;;
let binheap_to_seq t =
  List.rev (binheap_to_seq_ t [])
;;

(* 0 ... n - 1 *)
let rec range_ i n xs =
  if i = n then
    xs
  else
    range_ (i + 1) n (i :: xs)
;;
let range n =
  List.rev (range_ 0 n [])
;;

(* random sequence *)
let rec random_seq_ i n x seq =
  if i = n then
    seq
  else
    let y = (0x5DEECE66D * x + 0xB) land 0xFFFFFFFFFFFF in
    random_seq_ (i + 1) n y ((y lsr 17) :: seq)
;;
let rec random_seq n seed =
  List.rev (random_seq_ 0 n seed [])
;;

(* randomly shuffle xs *)
let random_shuffle seed xs =
  (List.combine (random_seq (List.length xs) seed) xs)
  |> List.sort (fun (a, _) (b, _) -> compare a b)
  |> List.map snd    
;;

(* print int list *)
let rec print_elems xs i n =
  if i = n then
    Printf.printf " ..."
  else
    match xs with
      [] -> ()
    | x :: rs ->
       (if i > 0 then
          Printf.printf ", %d" x
        else
          Printf.printf "%d" x;
        print_elems rs (i + 1) n)
;;

let print_list header xs n =
  Printf.printf "%s[" header;
  print_elems xs 0 n;
  Printf.printf "]\n";
  flush stdout
;;

(* add xs to empty tree; remove all elements from the tree;
   check if they are sorted *)
let check_seq xs =
  print_list "input:  " xs 7;
  let t = add_seq xs Empty in
  let ys = binheap_to_seq t in
  print_list "output: " ys 7;
  ys = (List.sort (-) xs)
;;

(* randomly shuffle 0 .. n-1;
   add them to empty tree;
   remove all elements from the tree;
   check they are sorted *)
let check_random seed n =
  let rng = range n in
  let xs = random_shuffle seed rng in
  check_seq xs
;;

let main () =
  let seed = 12345 in
  assert(check_random seed 2);
  assert(check_random seed 3);
  assert(check_random seed 4);
  assert(check_random seed 5);
  assert(check_random seed 10);
  assert(check_random seed 100);
  assert(check_random seed 1000);
  assert(check_random seed 10000);
  assert(check_random seed 100000);
  Printf.printf "OK\n"
;;

main()
